### PR TITLE
Named arguments in commands + many grammar fixes

### LINKED ISSUE #442: Change command + `=` syntax
Currently this:

```nim
foo a = b
```

parses as this:

```nim
foo(a) = b
```

The proposal is to make this mean:

```nim
foo(a = b)
```

and maybe even allow `foo a = b, c = d` syntax.

Reasoning:

* I can't think of a justification for the current syntax. `foo(a) = b` or `a.foo = b` are already possible and not much less ergonomic.
* Other languages with command syntax, such as Ruby, already employ this proposed syntax, which just makes Nim's current syntax weirder.
* I've never seen the current syntax in code likely because of the above problems.
* Named arguments in Nim are used like `foo(a = b, c = d)`, so there is a clear use for this syntax in command form. Macros and DSLs would also benefit from this.

This change would definitely break code which uses the current syntax. However I genuinely have never seen *any* code which does. And it can be easily tied to a big, backwards-incompatible release like 2.0.

### LINKED ISSUE #8846: [major] `someObject is object|int|tuple` fails (but `someObject is int|object|tuple` works...)
/cc @araq
reduced from a real use case:
```nim
type ObjectLike4* = tuple|object|int
type ObjectLike5* = int|tuple|object
type ObjectLike6* = object|int|tuple

type A = object
  tint: int

echo A is ObjectLike4
echo A is ObjectLike5
echo A is ObjectLike6

echo (1,2) is ObjectLike4
echo (1,2) is ObjectLike5
echo (1,2) is ObjectLike6

echo () is ObjectLike4
echo () is ObjectLike5
echo () is ObjectLike6

var a:A
echo a is ObjectLike4
echo a is ObjectLike5
echo a is ObjectLike6
```

prints:
true
true
false
false
true
true
true
true
true
true
true
false

every entry marked as false is a bug.

looks like:
* when T is object or tuple, `T is Foo1|Foo2|Foo3` doesn't take into account the 1st element Foo`
* when T is int, no bug
* when T is an instance of a tuple, there is a bug IF the instance is of size > 0

seems like a pretty serious bug...

## probably related
```nim
#BUG:D20180902T180936:here
type ObjectLike = object|tuple
when defined(case1):
  # Error: type mismatch: got <A>
  proc fun*[T:ObjectLike](a:T)=discard
when defined(case2):
  # ok
  proc fun*[T:tuple|object](a:T)=discard

type A = object
fun(A())
```

### LINKED ISSUE #15050: Command calls don't work with a single named arg
### Example
```nim
proc myProc(x = 0) = discard

myProc x=1 # error: undeclared identifier: 'x'
```

### Additional Information
* 1.2.4, 14d16c2174a4fcefd405f8c4eeaf57a6dcd2731d

### PR DESCRIPTION
**Pull request by metagn**
closes https://github.com/nim-lang/RFCs/issues/442, closes https://github.com/nim-lang/Nim/issues/15050, closes https://github.com/nim-lang/Nim/issues/8846 and maybe some other issues, cleans up after https://github.com/nim-lang/Nim/pull/19181

This is a fully rebased version of #19383 I am opening again to bump it up because apparently the time is ticking on 2.0 and it doesn't make sense for it to not have this.

`foo a = b` now means `foo(a = b)` instead of `foo(a) = b`, for consistency with `foo a, b = c`. To do this, the `primary` expression in the grammar, which had many issues and did not reflect reality, had to be slightly overhauled.

* Type expressions (including nominal type declarations!) are not directly part of `primary` anymore. Instead they have been broken up into `typeDefValue` (right hand sides of type sections) and `typeDesc` (everywhere else a type is expected).
* `typeDesc` itself is broken up into 2: `rawTypeDesc` (expressions that start with type keywords) and `typeDescExpr` (general expressions that may evaluate to types, as well as binary `not`). We use `rawTypeDesc` as the ordered choice in `primary`.
* Sigil operators are now in the grammar and only apply to `identOrLiteral`s for simplicity. (This seems to break nothing as it doesn't really make sense to apply to anything else)
* Command syntax is not part of `primarySuffix` anymore. This never made sense and had many issues. For example, in reality `primary` only parses 1 command argument, but the grammar denotes an arbitrary number of arguments.
* Full command syntax (i.e. with multiple arguments and `=`) is now in `exprStmt` (which it already was, but not in the grammar). We indicate with a parser mode that we want to get away with parsing only a `simplePrimary` (which is also reflected in the grammar), and if we do we parse `exprEqExpr ^+ comma` for command syntax. Otherwise we just use the expression that we got.
* As a minor unrelated grammar fix, we document `for a, (b, c) in d` syntax with `'for' ((varTuple / identWithPragma) ^+ comma)` instead of `'for' (identWithPragma ^+ comma)`.

### CODE REVIEW

File: doc/grammar.txt [hunk 1]
```diff
@@ -60,25 +61,25 @@ primarySuffix = '(' (exprColonEqExpr comma?)* ')'
...
+                  (':' optInd typeDescExpr)? ('=' optInd expr)?
 identColonEquals = IDENT (comma IDENT)* comma?
-     (':' optInd typeDesc)? ('=' optInd expr)?)
-tupleDecl = 'tuple'
-    '[' optInd  (identColonEquals (comma/semicolon)?)*  optPar ']' |
-    COMMENT? (IND{>} identColonEquals (IND{=} identColonEquals)*)?
+     (':' optInd typeDescExpr)? ('=' optInd expr)?)
+tupleType = 'tuple' '[' optInd (identColonEquals (comma/semicolon)?)* optPar ']'
+tupleDecl = tupleType |
+    'tuple' COMMENT? (IND{>} identColonEquals (IND{=} identColonEquals)*)?
```
Comment (Araq) [inline doc/grammar.txt]: Overlapping first set: `tuple` cannot be disambiguated with an LL(1) infrastucture hence these rules cannot reflect reality.

In other words, it needs to be refactored to:

```suggestion
tupleTypeBracket = '[' optInd (identColonEquals (comma/semicolon)?)* optPar ']'
tupleDecl = 'tuple' (tupleTypeBracket |
    COMMENT? (IND{>} identColonEquals (IND{=} identColonEquals)*)?)
```

Comment (metagn) [reply doc/grammar.txt]: Done

File: compiler/parser.nim [hunk 2]
```diff
@@ -1199,8 +1200,12 @@ proc parseTypeDescKAux(p: var Parser, kind: TNodeKind,
   getTok(p)
   if p.tok.indent != -1 and p.tok.indent <= p.currInd: return
   optInd(p, result)
+  let isTypedef = mode == pmTypeDef and p.tok.tokType in {tkObject, tkTuple}
```
Comment (ringabout) [inline compiler/parser.nim:1203]: It seems that
```
type
  Named = tuple or int
```
doesn't work anymore. Is it expected?

Comment (metagn) [reply compiler/parser.nim:1203]: Yes, `tuple` in type sections can have a syntax similar to objects, so it becomes `tuple[]` which is misleading

Hence this bigger example doesn't compile in 1.6:

```nim
type Named = tuple or int

proc foo(x: Named) = echo x

foo((1, 2))
```

Comment (ringabout) [reply compiler/parser.nim:1203]: I see, thanks for your explanation.

**Pull request merged**